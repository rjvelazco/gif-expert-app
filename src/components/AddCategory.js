import React, { useRef, useState } from 'react';
import PropTypes from 'prop-types';


export const AddCategory = ({ setCategories }) => {
    const [inputValue, setInputValue] = useState('');
    const input = useRef(null);
    const handleInputChange = ({ target }) => {
        setInputValue(target.value);
    }
    const handleSubmit = (e) => {
        e.preventDefault();
        if(inputValue.trim().length) {
            setCategories((categories) => [inputValue, ...categories]);
            setInputValue('');
            input.current.value='';
        }
    }
    return (
        <form onSubmit={handleSubmit}>
            <input
                ref={input}
                type="text"
                placeholder="Add Category"
                onChange={handleInputChange}
            />
        </form>
    )
}


AddCategory.propTypes = {
    setCategories: PropTypes.func
};