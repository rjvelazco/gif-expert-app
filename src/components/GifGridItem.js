import React from 'react'

export const GifGridItem = ({category, title, url}) => {
    const repTitle =  `${category } - gif`.toUpperCase();
    return (
        <div className="card">
            <img src={url} alt={title} />
            <p>{ title ? title : repTitle }</p>
        </div>
    )
}
