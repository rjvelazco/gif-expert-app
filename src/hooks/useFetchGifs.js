import { useEffect, useState } from 'react';
import { getGifs } from '../helpers/getGifs';
export const useFetchGifs = (category) => {

    const [state, setState] = useState({
        data: [],
        loading: true
    });

    useEffect(() => {
        setState((state) => ({
            ...state,
            loading: false
        }));
        getGifs(category)
                .then((images) => {
                    setState({
                        data: images,
                        loading: false
                    })
                });
    }, []);

    return state;
}
