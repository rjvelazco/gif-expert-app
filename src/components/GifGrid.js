import React from 'react';
import PropTypes from 'prop-types';
import { GifGridItem } from './GifGridItem';
import { useFetchGifs } from '../hooks/useFetchGifs';

export const GifGrid = ({ category }) => {
    
    const { data:images, loading } = useFetchGifs(category);

    return (
        <>
            <h3 className="card-grid animate__animated animate__fadeIn">{ category }</h3>

            { loading && <p className="card-grid animate__animated flash">Loading</p> }

            <div className="card-grid animate__animated animate__fadeIn">
                { 
                    images.map((img) => (
                        <GifGridItem
                            key={img.id}
                            category={category}
                            {...img}
                        />
                        // img={img}
                    ))
                }
            </div>
        </>
    )
}


GifGrid.protoTypes = {
    category: PropTypes.string.isRequired
}