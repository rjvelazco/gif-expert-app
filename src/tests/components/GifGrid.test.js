import React from 'react';
import '@testing-library/jest-dom';
import { shallow } from 'enzyme';
import { GifGrid } from '../../components/GifGrid';
import { useFetchGifs } from '../../hooks/useFetchGifs';

jest.mock('../../hooks/useFetchGifs');

describe('Tests on <GifGrid />', () => {

    // useFetchGifs.mockReturnValue({ data: [], loading: true });

    beforeEach(() => {
        jest.clearAllMocks();
    });

    test('It should render <GifGrid /> component correctly', () => {
        useFetchGifs.mockReturnValue({ data: [], loading: true });
        const wrapper = shallow(<GifGrid category={"Smash"} />);
        expect( wrapper ).toMatchSnapshot();
    });

    test('It should show items when images are loaded using useFetchGifs', () => {
        const gifs = [{
            id: 'ABC',
            url: 'https://localhost/cualquier/cosa.jps',
            title: 'Cualquier cosa'
        }];
        useFetchGifs.mockReturnValue({ data: gifs, loading: false });

        const wrapper = shallow(<GifGrid category={"Smash"} />);

        // expect( wrapper ).toMatchSnapshot();
        expect( wrapper.find('p').exists() ).toBeFalsy();
        expect( wrapper.find('GifGridItem').length ).toBe( gifs.length );
    });

});