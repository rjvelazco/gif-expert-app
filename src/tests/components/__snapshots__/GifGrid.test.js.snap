// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests on <GifGrid /> It should render <GifGrid /> component correctly 1`] = `
<Fragment>
  <h3
    className="card-grid animate__animated animate__fadeIn"
  >
    Smash
  </h3>
  <p
    className="card-grid animate__animated flash"
  >
    Loading
  </p>
  <div
    className="card-grid animate__animated animate__fadeIn"
  />
</Fragment>
`;
